vyper/parser/context.py

```diff
@@ -104,28 +104,64 @@ def range_scope(self):
         yield
         self.in_range_expr = prev_value
 
-    def internal_memory_scope(self, scope_id):
-        # syntactic sugar for `make_blockscope` used to release
-        # memory after creating temporary internal variables
-        return self.make_blockscope(scope_id)
+    @contextlib.contextmanager
+    def internal_memory_scope(self):
+        """
+        Internal memory scope context manager.
+
+        Internal variables that are declared within this context are de-allocated
+        upon exitting the context.
+        """
+        scope_id = self._scope_id_iter
+        self._scope_id_iter += 1
+        self._scopes.add(scope_id)
+        yield
+
+        # Remove all variables that have specific scope_id attached
+        released = [
+            (k, v) for k, v in self.vars.items() if v.is_internal and scope_id in v.blockscopes
+        ]
+        for name, var in released:
+            self.memory_allocator.deallocate_memory(var.pos, var.size * 32)
+            del self.vars[name]
+
+        # Remove block scopes
+        self._scopes.remove(scope_id)
 
     @contextlib.contextmanager
-    def make_blockscope(self, blockscope_id):
-        self.blockscopes.add(blockscope_id)
+    def block_scope(self):
+        """
+        Block scope context manager.
+
+        All variables (public and internal) that are declared within this context
+        are de-allocated upon exiting the context.
+        """
+        scope_id = self._scope_id_iter
+        self._scope_id_iter += 1
+        self._scopes.add(scope_id)
         yield
 
-        # Remove all variables that have specific blockscope_id attached.
-        released = [(k, v) for k, v in self.vars.items() if blockscope_id in v.blockscopes]
+        # Remove all variables that have specific scope_id attached
+        released = [(k, v) for k, v in self.vars.items() if scope_id in v.blockscopes]
         for name, var in released:
             self.memory_allocator.deallocate_memory(var.pos, var.size * 32)
             del self.vars[name]
 
         # Remove block scopes
-        self.blockscopes.remove(blockscope_id)
+        self._scopes.remove(scope_id)
 
-    def _new_variable(self, name: str, typ: NodeType, var_pos: int) -> int:
+    def _new_variable(self, name: str, typ: NodeType, var_size: int, is_internal: bool) -> int:
+        if is_internal:
+            var_pos = self.memory_allocator.expand_memory(var_size)
+        else:
+            var_pos = self.memory_allocator.allocate_memory(var_size)
         self.vars[name] = VariableRecord(
-            name=name, pos=var_pos, typ=typ, mutable=True, blockscopes=self.blockscopes.copy(),
+            name=name,
+            pos=var_pos,
+            typ=typ,
+            mutable=True,
+            blockscopes=self._scopes.copy(),
```

Comment:
i'm not sure why a variable needs to have pointers to all parent scopes. this seems like it should just be a single scope_id.

Reply:
Good point.  I just copied this from existing code without thinking :grimacing: